File: samples/nodejs-typescript-food-catalog/src/functions/queue-content.ts
import { InvocationContext, app } from "@azure/functions";
import { ContentMessage, CrawlType, ItemAction } from "../common/ContentMessage";
import { config } from "../common/config";
import { client } from "../common/graphClient";
import { enqueueItemDeletion, enqueueItemUpdate } from "../common/queueClient";
import { addItemToTable, getItemIds, getLastModified, recordLastModified, removeItemFromTable } from "../common/tableClient";

const { notificationEndpoint: apiUrl } = config;

async function crawl(crawlType: CrawlType, context: InvocationContext) {
    switch (crawlType) {
        case 'full':
        case 'incremental':
            await crawlFullOrIncremental(crawlType, context);
            break;
        case 'removeDeleted':
            await removeDeleted(context);
            break;
    }
}

async function crawlFullOrIncremental(crawlType: CrawlType, context: InvocationContext) {
    let url = `${apiUrl}/api/products`;

    if (crawlType === 'incremental') {
        const lastModified = await getLastModified(context);
        url += `?$filter=last_modified_t gt ${lastModified}`;
    }

    context.log(`Retrieving items from ${url}...`);

    const res = await fetch(url);
    if (!res.ok) {
        context.log(`Error retrieving item from ${url}: ${res.statusText}`);
        return;
    }

    const products: Product[] = await res.json();
    context.log(`Retrieved ${products.length} items from ${url}`);

    for (const product of products) {
        context.log(`Enqueuing item update for ${product.id}...`);
        enqueueItemUpdate(product.id);
    }
}

async function removeDeleted(context: InvocationContext) {
    const url = `${apiUrl}/api/products`;

    context.log(`Retrieving items from ${url}...`);

    const res = await fetch(url);
    if (!res.ok) {
        context.log(`Error retrieving item from ${url}: ${res.statusText}`);
        return;
    }

    const products: Product[] = await res.json();
    context.log(`Retrieved ${products.length} items from ${url}`);

    context.log('Retrieving ingested items...');
    const ingestedItemIds = await getItemIds(context);

    ingestedItemIds.forEach(ingestedItemId => {
        if (products.find(product => product.id === ingestedItemId)) {
            context.log(`Item ${ingestedItemId} still exists, skipping...`);
        }
        else {
            context.log(`Item ${ingestedItemId} no longer exists, deleting...`);
            enqueueItemDeletion(ingestedItemId);
        }
    });
}

async function processItem(itemId: string, itemAction: ItemAction, context: InvocationContext) {
    switch (itemAction) {
        case 'update':
            await updateItem(itemId, context);
            break;
        case 'delete':
            await deleteItem(itemId, context);
            break;
    }
}

async function updateItem(itemId: string, context: InvocationContext) {
    const url = `${apiUrl}/api/products/${itemId}`;

    context.log(`Retrieving item from ${url}...`);

    const res = await fetch(url);
    if (!res.ok) {
        context.log(`Error retrieving item from ${url}: ${res.statusText}`);
        return;
    }

    const product: Product = await res.json();
    context.log(`Retrieved product from ${url}`);
    context.log(JSON.stringify(product, null, 2));

    const externalItem = {
        id: product.id,
        properties: {
            'categories@odata.type': 'Collection(String)',
            categories: product.categories?.replace(/en:/g, '')?.split(', ') ?? [''],
            ecoscore: product.ecoscore_grade,
            imageUrl: product.image_url,
            'ingredients@odata.type': 'Collection(String)',
            ingredients: product.ingredients_text?.split(', ') ?? [''],
            nutriscore: product.nutriscore_grade,
            'traces@odata.type': 'Collection(String)',
            traces: product.traces_tags?.replace(/en:/g, '')?.split(', ') ?? [''],
            name: product.product_name,
            url: product.url.replace('.net/', '.org/')
        },
        content: {
            value: product.product_name,
            type: 'text'
        },
        acl: [
            {
                accessType: 'grant',
                type: 'everyone',
                value: 'everyone'
            }
        ]
    }

    context.log(`Transformed item`);
    context.log(JSON.stringify(externalItem, null, 2));

    const externalItemUrl = `/external/connections/${config.connector.id}/items/${product.id}`;
    context.log(`Updating external item ${externalItemUrl}...`)

    await client
        .api(externalItemUrl)
        .header('content-type', 'application/json')
        .put(externalItem);

    context.log(`Adding item ${product.id} to table storage...`);
    // track item to support deletion
    await addItemToTable(product.id, context);
    context.log(`Tracking last modified date ${product.last_modified_t}`);
    // track last modified date for incremental crawl
    await recordLastModified(product.last_modified_t, context);
}

async function deleteItem(itemId: string, context: InvocationContext) {
    const externalItemUrl = `/external/connections/${config.connector.id}/items/${itemId}`;
    context.log(`Deleting external item ${externalItemUrl}...`)

    await client
        .api(externalItemUrl)
        .delete();

    context.log(`Removing item ${itemId} from table storage...`);
    await removeItemFromTable(itemId, context);
}

app.storageQueue("contentQueue", {
    connection: "AzureWebJobsStorage",
    queueName: "queue-content",
    handler: async (message: ContentMessage, context: InvocationContext) => {
        context.log('Received message from queue queue-content');
        context.log(JSON.stringify(message, null, 2));

        const { action, crawlType, itemAction, itemId } = message;

        switch (action) {
            case 'crawl':
                await crawl(crawlType, context);
                break;
            case 'item':
                await processItem(itemId, itemAction, context);
                break;
            default:
                break;
        }
    }
});

File: samples/dotnet-csharp-graphdocs-ttk/Functions/QueueContent.cs
using System.Diagnostics;
using System.Text.Json;
using Azure.Data.Tables;
using Azure.Storage.Queues;
using Azure.Storage.Queues.Models;
using GraphDocsConnector.Messages;
using Markdig;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Extensions.Logging;
using Microsoft.Graph;
using Microsoft.Graph.Models.ExternalConnectors;
using YamlDotNet.Serialization;

namespace GraphDocsConnector.Functions
{
    class DocsArticle : IMarkdown
    {
        [YamlMember(Alias = "title")]
        public string? Title { get; set; }
        [YamlMember(Alias = "description")]
        public string? Description { get; set; }
        public string? Markdown { get; set; }
        public string? Content { get; set; }
        public string? RelativePath { get; set; }
    }

    public class QueueContent
    {
        private readonly ILogger<QueueContent> _logger;
        private readonly GraphServiceClient _graphClient;
        private readonly HttpClient _documentsClient;
        private readonly QueueClient _queueContentClient;
        private readonly TableClient _tableExternalItemsClient;
        private readonly TableClient _tableStateClient;

        public QueueContent(GraphServiceClient graphServiceClient, DocumentsServiceClient documentServiceClient, QueueServiceClient queueClient, TableServiceClient tableClient, ILogger<QueueContent> logger)
        {
            _graphClient = graphServiceClient;
            _documentsClient = documentServiceClient.Client;
            _queueContentClient = queueClient.GetQueueClient("queue-content");
            _tableExternalItemsClient = tableClient.GetTableClient("externalitems");
            _tableStateClient = tableClient.GetTableClient("state");
            _logger = logger;
        }

        [Function(nameof(QueueContent))]
        public async Task Run([QueueTrigger("queue-content", Connection = "AzureWebJobsStorage")] QueueMessage message)
        {
            var contentMessage = JsonSerializer.Deserialize<ContentMessage>(message.MessageText);
            if (contentMessage is null)
            {
                _logger.LogError($"Couldn't deserialize message: {message.MessageText}");
                return;
            }

            Debug.Assert(contentMessage.Action is not null);

            switch (contentMessage.Action)
            {
                case ContentAction.Crawl:
                    await Crawl(contentMessage.CrawlType);
                    break;
                case ContentAction.Item:
                    await ProcessItem(contentMessage.ItemId, contentMessage.ItemAction);
                    break;
            }
        }

        private async Task ProcessItem(string? itemId, ItemAction? itemAction)
        {
            if (itemAction is null)
            {
                _logger.LogError("itemAction is null");
                return;
            }

            if (itemId is null)
            {
                _logger.LogError("itemId is null");
                return;
            }

            switch (itemAction)
            {
                case ItemAction.Update:
                    await UpdateItem(itemId);
                    break;
                case ItemAction.Delete:
                    await DeleteItem(itemId);
                    break;
            }
        }

        private async Task DeleteItem(string itemId)
        {
            _logger.LogInformation($"Deleting item {itemId}...");
            await _graphClient.External
                .Connections[ConnectionConfiguration.ExternalConnection.Id]
                .Items[itemId]
                .DeleteAsync();
        }

        private async Task UpdateItem(string itemId)
        {
            var url = $"{Environment.GetEnvironmentVariable("DocumentsApi:Url")}/documents/{itemId}";

            _logger.LogInformation($"Retrieving item from {url}...");

            FileInfo? file = null;

            try
            {
                var res = await _documentsClient.GetStringAsync(url);
                file = JsonSerializer.Deserialize<FileInfo>(res, Utils.JsonSerializerOptions);
                if (file is null)
                {
                    _logger.LogWarning("Received null response");
                    return;
                }

            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "An error has occurred while retrieving items");
                return;
            }

            _logger.LogDebug($"Retrieved item from {url}");
            _logger.LogDebug(JsonSerializer.Serialize(file));

            if (file.Id is null)
            {
                _logger.LogWarning("File ID is null");
                return;
            }

            if (file.Contents is null)
            {
                _logger.LogWarning("File has no content");
                return;
            }

            var baseUrl = new Uri("https://learn.microsoft.com/graph/");

            var doc = file.Contents.GetContents<DocsArticle>();
            doc.Content = Markdown.ToHtml(doc.Markdown ?? "");

            var externalItem = new ExternalItem
            {
                Id = file.Id,
                Properties = new()
                {
                    AdditionalData = new Dictionary<string, object>
                    {
                        { "title", doc.Title ?? "" },
                        { "description", doc.Description ?? "" },
                        { "url", new Uri(baseUrl, file.Id.Replace("__", "/").Replace(".md", "")).ToString() }
                    }
                },
                Content = new()
                {
                    Value = doc.Content ?? "",
                    Type = ExternalItemContentType.Html
                },
                Acl = new()
                {
                    new()
                    {
                        Type = AclType.Everyone,
                        Value = "everyone",
                        AccessType = AccessType.Grant
                    }
                }
            };

            _logger.LogInformation($"Loading item {externalItem.Id}...");
            await _graphClient.External
                .Connections[ConnectionConfiguration.ExternalConnection.Id]
                .Items[externalItem.Id]
                .PutAsync(externalItem);

            _logger.LogInformation($"Adding item {externalItem.Id} to table storage...");
            Table.AddItem(_tableExternalItemsClient, externalItem.Id);

            _logger.LogInformation($"Tracking last modified date {file.LastModified}...");
            Table.RecordLastModified(_tableStateClient, file.LastModified, _logger);
        }

        private async Task Crawl(CrawlType? crawlType)
        {
            if (crawlType is null)
            {
                _logger.LogError("crawlType is null");
                return;
            }

            switch (crawlType)
            {
                case CrawlType.Full:
                case CrawlType.Incremental:
                    await CrawlFullOrIncremental(crawlType);
                    break;
                case CrawlType.RemoveDeleted:
                    await RemoveDeleted();
                    break;
            }
        }

        private async Task RemoveDeleted()
        {
            var url = $"{Environment.GetEnvironmentVariable("DocumentsApi:Url")}/documents";

            try
            {
                var res = await _documentsClient.GetStringAsync(url);
                var files = JsonSerializer.Deserialize<FileInfo[]>(res, Utils.JsonSerializerOptions);
                if (files is null)
                {
                    _logger.LogWarning("Received null response");
                    return;
                }

                _logger.LogInformation($"Retrieved {files.Length} items from {url}");

                var ingestedItemIds = Table.GetItemIds(_tableExternalItemsClient);

                foreach (var ingestedItemId in ingestedItemIds)
                {
                    if (files.Any(f => f.Id == ingestedItemId))
                    {
                        _logger.LogInformation($"Item {ingestedItemId} still exists, skipping...");
                    }
                    else
                    {
                        _logger.LogInformation($"Item {ingestedItemId} no longer exists, deleting...");
                        Queue.EnqueueItemDeletion(_queueContentClient, ingestedItemId);
                    }
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "An error has occurred while retrieving items");
            }
        }

        private async Task CrawlFullOrIncremental(CrawlType? crawlType)
        {
            var url = $"{Environment.GetEnvironmentVariable("DocumentsApi:Url")}/documents";

            if (crawlType == CrawlType.Incremental)
            {
                var lastModified = Table.GetLastModified(_tableStateClient);
                url += $"?$filter=lastModified gt {lastModified}";
            }

            _logger.LogInformation($"Retrieving items from {url}...");

            try
            {
                var res = await _documentsClient.GetStringAsync(url);
                var files = JsonSerializer.Deserialize<FileInfo[]>(res, Utils.JsonSerializerOptions);
                if (files is null)
                {
                    _logger.LogWarning("Received null response");
                    return;
                }

                _logger.LogInformation($"Retrieved {files.Length} items from {url}");
                foreach (var file in files)
                {
                    if (file.Id is null)
                    {
                        _logger.LogWarning($"ID is null. Skipping...");
                        continue;
                    }

                    _logger.LogInformation($"Enqueueing item update for {file.Id}...");
                    Queue.EnqueueItemUpdate(_queueContentClient, file.Id);
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "An error has occurred while retrieving items");
            }
        }
    }
}
